### ISSUE TITLE
Running jobs in a serial queue?

### ISSUE THREAD

**Issue by PaulAik** [HangfireIO/Hangfire#245]
Is there any way I can ensure jobs are run serially as opposed to in parallel (even better - have a serial queue)? E.g. I'd like to ensure only one job of type X can be running at any given time.

I've reduced my number of workers down to 1, not sure if that'll do the trick for now? Does the worker count always get used even if I use BackgroundJob.Enqueue?

**Comment by dotnetdan** [comment]
Just to confirm, does [DisableConcurrentExecution] guarantee that jobs will be processed in the order they were enqueued?  In the event that one of the jobs on the queue fails with retry disabled, it will continue processing the others in order?

**Comment by dukefama** [comment]
Hi, I am interested in the response to dotnetdan's question. Is there anyone that can assist with this?

**Comment by JarrodJ83** [comment]
@dotnetdan I am seeing the exact opposite, actually. It seems the last in is the first processed. In below screenshot the "complete" job was the last thing I enqueued but it was the first thing that ran:

![image](https://cloud.githubusercontent.com/assets/248166/20186147/d1bdf5fe-a73c-11e6-9afe-4e78d7b3b4a4.png)

I would think this is a bug. Would like to hear @odinserj 's feedback, though.

**Comment by RudeySH** [comment]
Any update on this? After reading this thread I can't help but feel Hangfire queues are pretty useless. Is it still not possible to stop concurrent execution of different kinds of jobs? And was the queue order bug ever fixed?

**Comment by nmurf** [comment]
@odinserj, would you please respond to @JarrodJ83?

**Comment by BalassaMarton** [comment]
This is something we also need: we use HF in an event sourcing scenario where it is required that
- enqueued jobs (of the same method) run in strict FIFO order
- they are not processed in parallel, ever.

**Comment by brandlink-james** [comment]
Is there any update for this ?

**Comment by dazbradbury** [comment]
> This way works for me.
> app.UseHangfireServer(options: new BackgroundJobServerOptions { ServerName = "SequentialServer", WorkerCount = 1 });

This is great unless you have multiple hangfire servers. At that point, you need a way to ensure a single worker queue in a distributed environment, across deploys etc. Anyone have an elegant solution to that?

**Comment by Sire** [comment]
Check out their paid version, it might solve this: https://www.hangfire.io/ace/

**Comment by SoftCircuits** [comment]
`[DisableConcurrentExecution]`

> Error CS1729 'DisableConcurrentExecutionAttribute' does not contain a constructor that takes 0 arguments

**Comment by auvobws** [comment]
Did anyone figure out how to get the jobs to stay in enqueued until the first job is completely done: either failed/suceeded and not scheduled?

**Comment by DMiradakis** [comment]
I don't think this is offered in Hangfire, neither in the free version nor in the paid versions.

As far as I can tell, the "type" of queue used in Hangfire is what I've heard called a **Heap Queue**. In AWS Simple Queue Service, they call it a **Standard Queue**. The idea is that jobs are executed in a *best-effort ordering* in a highly concurrent manner, where exact ordering is NOT guaranteed and instead throughput is prioritized.

It sounds like what many of you are looking for is a FIFO queue, where strict ordering is **guaranteed** and throughput is deprioritized.

You'll probably need to open a PR and design this yourself, or use another service. I'm building an open source .NET job orchestrator called [Didact](https://www.didact.dev), and I'll be including both Heap Queues and FIFO Queues. Pretty soon I'll have some docs on this on the [docsite](https://docs.didact.dev).
